code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +160,221 @@ public void updateAccount(Account account, String name, String kind, String exte
+        }
+        projectMembers.addAll(create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch());
+        return !projectMembers.isEmpty();
     }
 
-    @Inject
-    public void setObjectManager(ObjectManager objectManager) {
-        this.objectManager = objectManager;
+    @Override
+    public Account updateProject(Account project, String name, String description) {
+        Map<TableField<AccountRecord, String>, String> properties = new HashMap<>();
+        properties.put(ACCOUNT.NAME, name);
+        properties.put(ACCOUNT.DESCRIPTION, description);
+        int updateCount = create()
+                .update(ACCOUNT)
+                .set(properties)
+                .where(ACCOUNT.ID
+                        .eq(project.getId()))
+                .execute();
+        if(1 != updateCount) {
+            throw new RuntimeException("UpdateProject failed.");
+        }
+        return getAccountById(project.getId());
+    }
+
+    public List<? extends ProjectMember> setProjectMembers(Account project, HashSet<Member> members){
+        List<? extends ProjectMember> previousMembers = getProjectMembers(project.getId());
+        HashSet<Member> otherPreviosMembers = new HashSet<>();
+        for (ProjectMember member: previousMembers){
+            otherPreviosMembers.add(new Member(member));
+        }
+        HashSet<Member> update = new HashSet<>();
+        HashSet<Member> createIgnore = new HashSet<>();
+        HashSet<Member> deleteIgnore = new HashSet<>();
+        for (Member member: otherPreviosMembers){
+            for (Member memberNew: members){
+                if (memberNew.getExternalId().equalsIgnoreCase(member.getExternalId()) &&
+                        memberNew.getExternalIdType().equalsIgnoreCase(member.getExternalIdType())) {
+                    if (!memberNew.getRole().equalsIgnoreCase(member.getRole())) {
+                        update.add(memberNew);
+                        deleteIgnore.add(member);
+                    } else {
+                        createIgnore.add(memberNew);
+                        deleteIgnore.add(member);
+                    }
+                }
+            }
+        }
+        HashSet<Member> create = (HashSet) members.clone();
+        create.removeAll(update);
+        create.removeAll(createIgnore);
+        HashSet<Member> delete = (HashSet) otherPreviosMembers.clone();
+        delete.removeAll(deleteIgnore);
+
+        Condition allMembers = DSL.falseCondition();
+        for (Member member: delete){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                    .and(PROJECT_MEMBER.PROJECT_ID.eq(project.getId())));
+        }
+        List<? extends ProjectMember> toDelete = create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch();
+        for (ProjectMember member: toDelete){
```

Comment:
There is a method somewhere in objectManager called deactivateThenRemove. I think its better to replace 2 scheduled processes below with this one.